crates/rethnet_rpc_server/src/lib.rs

```diff
@@ -0,0 +1,526 @@
+use hashbrown::HashMap;
+use rethnet_eth::remote::ZeroXPrefixedBytes;
+use tokio::sync::{Mutex, RwLock};
+
+use rethnet_eth::{
+    remote::{
+        client::{Request as RpcRequest, RpcClient},
+        jsonrpc,
+        jsonrpc::{Response, ResponseData},
+        methods::MethodInvocation as EthMethodInvocation,
+        BlockSpec,
+    },
+    Address, B256, U256,
+};
+use rethnet_evm::{
+    state::{AccountModifierFn, ForkState, HybridState, StateError, SyncState},
+    AccountInfo, Bytecode, KECCAK_EMPTY,
+};
+
+mod hardhat_methods;
+pub use hardhat_methods::{reset::RpcHardhatNetworkConfig, HardhatMethodInvocation};
+
+/// an RPC method with its parameters
+#[derive(Clone, Debug, PartialEq, serde::Deserialize, serde::Serialize)]
+#[serde(untagged)]
+#[allow(clippy::large_enum_variant)]
+pub enum MethodInvocation {
+    /// an eth_* method invocation
+    Eth(EthMethodInvocation),
+    /// a hardhat_* method invocation
+    Hardhat(HardhatMethodInvocation),
+}
+
+type RethnetStateType = Arc<RwLock<Box<dyn SyncState<StateError>>>>;
+
+struct AppState {
+    rethnet_state: RethnetStateType,
+    fork_client: Option<Arc<Mutex<RpcClient>>>,
+    fork_block_number: Option<usize>,
+}
+
+type StateType = Arc<AppState>;
+
+fn response<T>(id: jsonrpc::Id, data: ResponseData<T>) -> (StatusCode, Json<serde_json::Value>)
+where
+    T: serde::Serialize,
+{
+    let response: Response<T> = Response {
+        jsonrpc: jsonrpc::Version::V2_0,
+        id: id.clone(),
+        data,
+    };
+    if let Ok(response) = serde_json::to_value(response) {
+        (StatusCode::OK, Json(response))
+    } else {
+        (
+            StatusCode::INTERNAL_SERVER_ERROR,
+            Json(
+                serde_json::to_value(Response {
+                    jsonrpc: jsonrpc::Version::V2_0,
+                    id,
+                    data: ResponseData::<T>::new_error(0, "serde_json::to_value() failed", None),
+                })
+                .expect("shouldn't happen"),
+            ),
+        )
+    }
+}
+
+/// returns the state root in effect BEFORE setting the block context, so that the caller can
+/// restore the context to that state root.
+async fn set_block_context<T>(
+    state: &StateType,
+    block_spec: BlockSpec,
+) -> Result<B256, ResponseData<T>> {
+    let previous_state_root =
+        state.rethnet_state.read().await.state_root().map_err(|_| {
+            ResponseData::new_error(0, "Failed to retrieve previous state root", None)
+        })?;
+    if block_spec == BlockSpec::Tag(String::from("latest")) {
+        // do nothing
+        Ok(previous_state_root)
+    } else {
+        state
+            .rethnet_state
+            .write()
+            .await
+            .set_block_context(
+                &KECCAK_EMPTY,
+                Some(match block_spec {
+                    BlockSpec::Number(n) => Ok(n),
+                    BlockSpec::Tag(tag) => {
+                        if let Some(fork_client) = state.fork_client.clone() {
```

Comment:
This is still not the same logic as currently exists in Hardhat: https://github.com/NomicFoundation/hardhat/blob/main/packages/hardhat-core/src/internal/hardhat-network/provider/modules/eth.ts#L1388

So far, all of our code is a port from Hardhat's logic to Rust. As such, we need to make sure that we replicate all logic. When writing code, the ground truth should be the TypeScript code. The Hardhat tests that we run and validate afterwards merely serve as confirmation that we did it correctly.

If you haven't done this yet, I'd recommend validating that all of implementations actually follow the same logic as the [Eth](https://github.com/NomicFoundation/hardhat/blob/main/packages/hardhat-core/src/internal/hardhat-network/provider/modules/eth.ts) and [Hardhat](https://github.com/NomicFoundation/hardhat/blob/main/packages/hardhat-core/src/internal/hardhat-network/provider/modules/hardhat.ts) modules.

Reply:
Addressed in https://github.com/NomicFoundation/hardhat/pull/3998/commits/2cda55e1d9064f13c00e607ec7558131f7c08dad